### PR TITLE
S8 restrict api

### PR DESCRIPTION
**Pull request by PapaPwn**
Restrict the backend too, plus some more login redirection, plus a couple of bug fixes.
AND
Make the frontend respond with prompts to login.

### CODE REVIEW

File: backend/package.json [hunk 1]
```diff
@@ -16,6 +16,7 @@
     "express": "~4.16.0",
     "firebase-admin": "^6.0.0",
     "morgan": "~1.9.0",
+    "node-fetch": "^2.2.0",
```
Comment (DarkPurple141) [inline backend/package.json]: This is a dev dependency, not a application dependency. We only use it for testing.

File: backend/src/utils/cors.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ exports.cors = function(req, res, next) {
 }
 
 exports.corsProd = function(req, res, next) {
-    res.header('Access-Control-Allow-Origin', 'https://smartcourse.me')
+    res.header('Access-Control-Allow-Origin', 'https://smartcourse.me, https://wwww.smartcourse.me')
```
Comment (DarkPurple141) [inline backend/src/utils/cors.js]: I'm fairly sure you can't specify multiple origin in this header. The only way to deal with this is to dynamically update the header at runtime depending on the request.

File: frontend/src/views/Review.vue [hunk 3]
```diff
@@ -53,6 +53,12 @@ export default {
   },
   methods: {
     submitReply (replyForm) {
+      const authState = this.$store.getters['auth/isLoggedIn']
```
Comment (DarkPurple141) [inline frontend/src/views/Review.vue]: Don't do this. We're going to remove the functionality. And prompt, not force them.

Comment (DarkPurple141) [reply frontend/src/views/Review.vue]: Same with above.

File: frontend/src/components/reviews-replies/ReviewCard.vue [hunk 4]
```diff
@@ -37,11 +37,21 @@ export default {
   },
   methods: {
     upvote() {
+      const authState = this.$store.getters['auth/isLoggedIn']
```
Comment (DarkPurple141) [inline frontend/src/components/reviews-replies/ReviewCard.vue]: Same with this. I think better to prompt user then force redirect.

File: backend/src/utils/cors.js (hunk 2 above)
Comment (PapaPwn) [reply backend/src/utils/cors.js]: Yeah you are right, will fix this.

File: frontend/src/views/Review.vue (hunk 3 above)
Comment (PapaPwn) [reply frontend/src/views/Review.vue]: Do we have hidden prompt functionality or should I implement this now?

Comment (DarkPurple141) [reply frontend/src/views/Review.vue]: There's not presently. I was thinking button would be replaced with text if user is not authenticated. "You need to be logged in to write a review" etc OR button goes to disabled state and has a hover text.

Comment (PapaPwn) [reply frontend/src/views/Review.vue]: That's a lot of text for a button. The second option might be viable. I'll have a play and see what feels natural :)

Comment (DarkPurple141) [reply frontend/src/views/Review.vue]: It wouldn't be a button.

Comment (NunoDasNeves) [reply frontend/src/views/Review.vue]: ^

File: frontend/src/views/Question.vue [hunk 5]
```diff
@@ -1,16 +1,16 @@
 <template>
     <section class="main-content">
       <AppBreadCrumb/>
-      <QuestionCard v-bind="question"/>
+      <QuestionCard v-bind="question" :authenticated="authenticated"/>
 
-      <AnswerForm @submitCommentForm="submitAnswer" :type="commentType" :callback="submitAnswer">
+      <AnswerForm @submitCommentForm="submitAnswer" :type="commentType" :callback="submitAnswer" :authenticated="authenticated">
```
Comment (NunoDasNeves) [inline frontend/src/views/Question.vue]: For the forms I think it'd be better if we replaced the whole form with text (e.g. "Login to answer this question") or ifwe greyed out/disabled the whole form

Comment (DarkPurple141) [reply frontend/src/views/Question.vue]: Agree with the first part. I think the former just disappears and becomes a sentence of text.

Comment (PapaPwn) [reply frontend/src/views/Question.vue]: No problemo, but making things pretty isn't a strength of mine, so I might leave the CSS to @DarkPurple141 , but the functionality will be there.

Comment (PapaPwn) [reply frontend/src/views/Question.vue]: Actually the functionality is already there you just need to add `v-if="authenticated"` and a `v-else` :)

**Comment by DarkPurple141** [comment]
I'll be interested to know how much that v-tooltip adds to the bundle size.

File: frontend/src/components/AppButton.vue [hunk 6]
```diff
@@ -1,7 +1,16 @@
 <template>
-    <button type='button' :class="alt ? 'standard-button' : 'alt-button'">
-        <slot></slot>
-    </button>
+    <div>
```
Comment (DarkPurple141) [inline frontend/src/components/AppButton.vue]: This seems like two components.  Else, one component but oddly shaped.

File: frontend/src/components/Comments/CommentForm.vue [hunk 7]
```diff
@@ -2,7 +2,9 @@
     <CardForm :title="`Submit ${type}`">
       <textarea placeholder="Your input here.." v-model="body"></textarea><br>
       <!-- should probs be a separate component -->
-      <AppButton @click.native="callback({body}); body=''">{{ type }}</AppButton>
+      <AppButton :disabled="!(authenticated)" :disabledMsg="disabledMsg" @click.native="callback({body}); body=''">
```
Comment (DarkPurple141) [inline frontend/src/components/Comments/CommentForm.vue]: brackets?

Comment (PapaPwn) [reply frontend/src/components/Comments/CommentForm.vue]: No reason Mr Linter.

File: frontend/src/components/AppButton.vue (hunk 6 above)
Comment (PapaPwn) [reply frontend/src/components/AppButton.vue]: It's one and oddly shaped, just wanted something that worked. Not sure about the proper pattern here.

**Comment by PapaPwn** [comment]
Going to merge this in and the style changes can be made as a separate PR.

**Pull request merged**